src/os/bluestore/BlueStore.cc

```diff
@@ -7471,28 +7472,31 @@ int BlueStore::_do_write(
   wctx.csum_order = block_size_order;
 
   // compression parameters
-  unsigned alloc_hints = o->onode.alloc_hint_flags;
-  wctx.compress =
-    (comp_mode == COMP_FORCE) ||
-    (comp_mode == COMP_AGGRESSIVE &&
-     (alloc_hints & CEPH_OSD_ALLOC_HINT_FLAG_INCOMPRESSIBLE) == 0) ||
-    (comp_mode == COMP_PASSIVE &&
-     (alloc_hints & CEPH_OSD_ALLOC_HINT_FLAG_COMPRESSIBLE));
-
-  if ((alloc_hints & CEPH_OSD_ALLOC_HINT_FLAG_SEQUENTIAL_READ) &&
-      (alloc_hints & CEPH_OSD_ALLOC_HINT_FLAG_RANDOM_READ) == 0 &&
-      (alloc_hints & (CEPH_OSD_ALLOC_HINT_FLAG_IMMUTABLE|
+  {
+    std::lock_guard<std::mutex> l(comp_lock);
+    unsigned alloc_hints = o->onode.alloc_hint_flags;
+    wctx.compress =
+      (comp_mode == COMP_FORCE) ||
+      (comp_mode == COMP_AGGRESSIVE &&
+       (alloc_hints & CEPH_OSD_ALLOC_HINT_FLAG_INCOMPRESSIBLE) == 0) ||
+      (comp_mode == COMP_PASSIVE &&
+       (alloc_hints & CEPH_OSD_ALLOC_HINT_FLAG_COMPRESSIBLE));
+
+    if ((alloc_hints & CEPH_OSD_ALLOC_HINT_FLAG_SEQUENTIAL_READ) &&
+        (alloc_hints & CEPH_OSD_ALLOC_HINT_FLAG_RANDOM_READ) == 0 &&
+        (alloc_hints & (CEPH_OSD_ALLOC_HINT_FLAG_IMMUTABLE|
 			CEPH_OSD_ALLOC_HINT_FLAG_APPEND_ONLY)) &&
-      (alloc_hints & CEPH_OSD_ALLOC_HINT_FLAG_RANDOM_WRITE) == 0) {
-    dout(20) << __func__ << " will prefer large blob and csum sizes" << dendl;
-    wctx.comp_blob_size = comp_max_blob_size;
-    wctx.csum_order = min_alloc_size_order;
-  } else {
-    wctx.comp_blob_size = comp_min_blob_size;
+        (alloc_hints & CEPH_OSD_ALLOC_HINT_FLAG_RANDOM_WRITE) == 0) {
+      dout(20) << __func__ << " will prefer large blob and csum sizes" << dendl;
+      wctx.comp_blob_size = comp_max_blob_size;
+      wctx.csum_order = min_alloc_size_order;
+    } else {
+      wctx.comp_blob_size = comp_min_blob_size;
+    }
+    dout(20) << __func__ << " prefer csum_order " << wctx.csum_order
+	     << " comp_blob_size 0x" << std::hex << wctx.comp_blob_size
+	     << std::dec << dendl;
```

Comment:
I'm worried about the cost of this lock on every write op.  We are looking at 3 values, comp_mode, and the min and max blob sizes.  They are all a single work.  I don't think it matters if they are updated incoherently (e.g., new max blob  size, old comp_mode).  Is there a specific case you're worried about?  Can we solve it with atomics instead of a lock?